core/internal/runmetric/runmetric.go

```diff
@@ -3,67 +3,201 @@ package runmetric
 import (
 	"errors"
 	"path/filepath"
+	"strings"
 
+	"github.com/wandb/wandb/core/internal/pathtree"
+	"github.com/wandb/wandb/core/internal/runhistory"
+	"github.com/wandb/wandb/core/internal/runsummary"
 	"github.com/wandb/wandb/core/pkg/service"
 	"google.golang.org/protobuf/proto"
 )
 
 type MetricHandler struct {
-	DefinedMetrics map[string]*service.MetricRecord
-	GlobMetrics    map[string]*service.MetricRecord
+	metricDefs     map[string]*service.MetricRecord
+	globMetricDefs map[string]*service.MetricRecord
 }
 
 func NewMetricHandler() *MetricHandler {
 	return &MetricHandler{
-		DefinedMetrics: make(map[string]*service.MetricRecord),
-		GlobMetrics:    make(map[string]*service.MetricRecord),
+		metricDefs:     make(map[string]*service.MetricRecord),
+		globMetricDefs: make(map[string]*service.MetricRecord),
 	}
 }
 
-// createMatchingGlobMetric check if a key matches a glob pattern, if it does create a new defined metric
-// based on the glob metric and return it.
-func (mh *MetricHandler) CreateMatchingGlobMetric(key string) *service.MetricRecord {
-	for pattern, globMetric := range mh.GlobMetrics {
-		if match, err := filepath.Match(pattern, key); err != nil {
+// Exists reports whether a non-glob metric is defined.
+func (mh *MetricHandler) Exists(key string) bool {
+	_, exists := mh.metricDefs[key]
+	return exists
+}
+
+// InsertStepMetrics inserts an automatic step metric for every defined
+// metric with step_sync set to true.
+//
+// TODO: Remove the 'summary' parameter from 'InsertStepMetrics'.
+// The 'summary' parameter is used by the old implementation, which is
+// to be refactored.
+func (mh *MetricHandler) InsertStepMetrics(
+	history *runhistory.RunHistory,
+	summary *runsummary.RunSummary,
+) {
+	for _, metricDef := range mh.metricDefs {
+		// Skip any metrics that do not need to be synced.
+		if metricDef.StepMetric == "" || !metricDef.Options.GetStepSync() {
+			continue
+		}
+
+		// Skip if the step is already set.
+		if history.Contains(metricDef.StepMetric) {
+			continue
+		}
+
+		// Skip if the metric doesn't show up in history.
+		if !history.Contains(metricDef.Name) {
 			continue
-		} else if match {
-			metric := proto.Clone(globMetric).(*service.MetricRecord)
-			metric.Name = key
-			metric.Options.Defined = false
-			metric.GlobName = ""
-			return metric
 		}
+
+		// TODO: Insert the latest value.
+		mh.hackInsertLatestValue(metricDef.StepMetric, history, summary)
+	}
+}
+
+func (mh *MetricHandler) hackInsertLatestValue(
+	stepMetricKey string,
+	history *runhistory.RunHistory,
+	summary *runsummary.RunSummary,
+) {
+	value, exists := summary.Get(stepMetricKey)
+	if !exists {
+		return
+	}
+
+	// NOTE: This assumes that the source value is always a float64 or int64
+	// if it is a number. This is true because all JSON libraries we use decode
+	// numbers into float64 or int64, but this is a brittle assumption.
```

Comment:
yes. That behavior is documented (see simplejsonext docs and encoding/json docs), but I'm refactoring this soon, so it will hopefully not stay in the codebase long.